bf16 project goal (#723)
rust-lang/rust-project-goals: src/2026/bf16.md

# `bf16` Primitive Type

| Metadata         |                                    |

> | :--------------- | :--------------------------------- |
> | Point of contact | @Jamesbarford | 
> | Status           | Proposed                           |
> 

Reviewer: ```suggestion
| Status           | Proposed                           |
| [compiler] champion | @davidtwco |

```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..2f874e5149c11ffca8475fb2077b7a8276752c1f)*

of the value. It also cannot use the usual floating-point surface of the
language, such as casts, literals, constants, formatting, arithmetic traits,
or generic bounds over floating-point types.

As such, workarounds are required to use a `bf16` type today.
[rust-lang/rust#140763] is an example of the current workaround: it adds
temporary codegen bypasses so Rust signatures using `i16` / `i16xN` can call
LLVM intrinsics whose real signatures use `bf16` / `bf16xN`.

### What we propose to do about it

We propose a primitive `bf16` type that can be used across architectures.

### Work items over the next year

| Task        | Owner(s) | Notes |
| ----------- | -------- | ----- |
| Draft RFC | *Jamesbarford*  | |
| Resolve feedback on the RFC | *Jamesbarford* | |
| Approval for an experiement behind a `!feature(bf16)` | *Jamesbarford* | |
| Implement experiment | *Jamesbarford* | |
| Stabilisation concerns | *Jamesbarford* | |
| Write documentation, references and limitations | *Jamesbarford* | |
| Stabilise | *Jamesbarford* | |

## Team asks

| Team | Support level | Notes |
|------|-----------------|-----------------|
| [lang] | Medium | Review of RFC |

> | [compiler] | Medium | PR Reviews, both of the RFC and of the subsequent implementation work |

Reviewer: Needs lang and compiler champions

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..2f874e5149c11ffca8475fb2077b7a8276752c1f)*

Reviewer 2: @davidtwco is compiler champion, per [the 2026-08-05 language team triage meeting](https://hackmd.io/cn7IRnfLT9eUh8pBsm0FdQ?view), @tmandry will be the lang champion

# `bf16` Primitive Type

| Metadata         |                                    |
| :--------------- | :--------------------------------- |
| Point of contact | @Jamesbarford | 
| Status           | Proposed                           |
| [compiler] champion | @davidtwco |

*Arm's Rust team is @davidtwco, @adamgemmell, @lqd and @Jamesbarford. This goal
will be primarily worked on by @Jamesbarford*

## Summary

Implementation of a `bf16` type in Rust.

## Motivation

A first-class `bf16` type is primarily useful for target-specific SIMD and
Matrix APIs. It is also useful where memory bandwidth, storage size, and vector

> throughput matter more than decimal precision. The main use case is machine

Reviewer 3: Lang hat: we have `f16` for these motivations as a primitive.

Does `bf16` *need* to be a primitive with full support for everything?  If the goal is "target-specific SIMD and Matrix APIs", then does it need to be a *full* primitive?  Could the goal be cut down to "we have a better type to use in intrinsics taking `bf16`" and punt on everything else?

(👍 lang wise for having the restricted form.  Doing the full version with I think I'm 👎 when we don't even have that for `f16` completely yet, and that shows how big the work is to *fully* support something.)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..3ce351ff7bd24803fd7b5a770f68b91c981d4f1e)*

Author: Re-worded to say;

> We propose a `bf16` type in `core::num` that can be used across architectures.

learning and AI workloads, where values are often stored or multiplied in
`bf16` while accumulation happens in wider precision.

`b16` is increasingly found in hardware across multiple architecutues such as
[`AArch64` Neoverse N2][neoverse_n2], [`x86_64` Cooper Lake][cooperlake] and
[`RISC-V`][riscv]. All offering a level of native support for scalar and vector
operations with discoverability through target-feature mechanisms.

### The status quo

Presently `bf16` is present through wrapper types over a `u16` with adhoc
lowering rules. Using a `u16` preserves the raw bits, but it loses the meaning
of the value. It also cannot use the usual floating-point surface of the
language, such as casts, literals, constants, formatting, arithmetic traits,
or generic bounds over floating-point types.

As such, workarounds are required to use a `bf16` type today.
[rust-lang/rust#140763] is an example of the current workaround: it adds
temporary codegen bypasses so Rust signatures using `i16` / `i16xN` can call
LLVM intrinsics whose real signatures use `bf16` / `bf16xN`.

### What we propose to do about it

> We propose a `bf16` type in `core::num` that can be used across architectures.

Reviewer 3: I would propose being more explicit about what "used" means here.

In the spirit of "with discoverability through target-feature mechanism" from earlier, maybe something like

```suggestion
We propose a `bf16` type in `core::num` that can be used across architectures
for FFI and with target-specific intrinsics.

*This* goal does not propose to make it a general-purpose arithmetic type,
so does not propose portable rust operations, portable library methods, etc.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..84f1f40debc49f7acd2fefbcd10e3596ca8017ab)*

Author: c0d3741d91df2d2ed457f37822edc35283ec6eff :+1:

## Summary

Implementation of a `bf16` type in Rust.

## Motivation

A first-class `bf16` type is primarily useful for target-specific SIMD and
Matrix APIs. It is also useful where memory bandwidth, storage size, and vector
throughput matter more than decimal precision. The main use case is machine
learning and AI workloads, where values are often stored or multiplied in
`bf16` while accumulation happens in wider precision.

`b16` is increasingly found in hardware across multiple architecutues such as
[`AArch64` Neoverse N2][neoverse_n2], [`x86_64` Cooper Lake][cooperlake] and
[`RISC-V`][riscv]. All offering a level of native support for scalar and vector
operations with discoverability through target-feature mechanisms.

### The status quo

Presently `bf16` is present through wrapper types over a `u16` with adhoc
lowering rules. Using a `u16` preserves the raw bits, but it loses the meaning
of the value. It also cannot use the usual floating-point surface of the
language, such as casts, literals, constants, formatting, arithmetic traits,
or generic bounds over floating-point types.

As such, workarounds are required to use a `bf16` type today.

> [PR #140763][pr_140763] is an example of the current workaround: it adds

Reviewer 4: the common patter to link prs or issues like this is `[rust-lang/rust#140763]` :) Then you could also skip the link further down, because those patterns get linked automatically.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..e56d40d83d5dae3506f2daece9da0c5ede074655)*

Reviewer: ```suggestion
[rust-lang/rust#140763] is an example of the current workaround: it adds
```

temporary codegen bypasses so Rust signatures using `i16` / `i16xN` can call
LLVM intrinsics whose real signatures use `bf16` / `bf16xN`.

### What we propose to do about it

We propose a primitive `bf16` type that can be used across architectures.

### Work items over the next year

| Task        | Owner(s) | Notes |
| ----------- | -------- | ----- |
| Draft RFC | *Jamesbarford*  | |
| Resolve feedback on the RFC | *Jamesbarford* | |
| Approval for an experiement behind a `!feature(bf16)` | *Jamesbarford* | |
| Implement experiment | *Jamesbarford* | |
| Stabilisation concerns | *Jamesbarford* | |
| Write documentation, references and limitations | *Jamesbarford* | |
| Stabilise | *Jamesbarford* | |

## Team asks

| Team | Support level | Notes |
|------|-----------------|-----------------|
| [lang][lang]| Medium | Review of RFC |

> | [compiler][compiler] | Medium | PR Reviews, both of the RFC and of the subsequent implementation work |

Reviewer: ```suggestion
| [lang] | Medium | Review of RFC |
| [compiler] | Medium | PR Reviews, both of the RFC and of the subsequent implementation work |
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..e56d40d83d5dae3506f2daece9da0c5ede074655)*

## Funding

The `bf16` is being worked on by the Rust team at Arm and the goal owners have
sufficient capacity to do the work.

## Frequently asked questions

- What does the `b` in `bf16` mean? The `b` stands for `brain`

[neoverse_n2]: https://developer.arm.com/documentation/102099/0003/The-Neoverse-N2--core/Supported-standards-and-specifications
[cooperlake]: https://en.wikipedia.org/wiki/Cooper_Lake_(microprocessor)
[riscv]: https://docs.riscv.org/reference/isa/unpriv/bfloat16.html#insns-fcvt.s.bf16
[pr_140763]: https://github.com/rust-lang/rust/pull/140763
[lang]: https://github.com/rust-lang/lang-team

> [compiler]: https://github.com/rust-lang/compiler-team

Reviewer: ```suggestion
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/723/a535ed930c8e6f0b3448f3eb96399ca9a5f1bb16..e56d40d83d5dae3506f2daece9da0c5ede074655)*
